src/paginate.ts

```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
+          }
+        : { propertyName: propertyPath[0] }
+}
+
+function extractVirtualProperty(
+    qb: SelectQueryBuilder<unknown>,
+    columnProperties: ColumnProperties
+): { isVirtualProperty: boolean; query?: ColumnMetadata['query'] } {
+    const metadata = columnProperties.propertyPath
+        ? qb?.expressionMap?.mainAlias?.metadata?.findColumnWithPropertyPath(columnProperties.propertyPath)
+              ?.referencedColumn?.entityMetadata // on relation
+        : qb?.expressionMap?.mainAlias?.metadata
+    return (
+        metadata?.columns?.find((column) => column.propertyName === columnProperties.propertyName) || {
+            isVirtualProperty: false,
+            query: undefined,
+        }
+    )
+}
+
+function checkIsRelation(qb: SelectQueryBuilder<unknown>, propertyPath: string): boolean {
+    if (!qb || !propertyPath) {
+        return false
+    }
+    return !!qb?.expressionMap?.mainAlias?.metadata?.hasRelationWithPropertyPath(propertyPath)
+}
+
+// This function is used to fix the query parameters when using relation, embeded or virtual properties
+// It will replace the column name with the alias name and return the new parameters
+function fixQueryParam(
+    alias: string,
+    column: string,
+    filter: Filter,
+    condition: WherePredicateOperator,
+    parameters: { [key: string]: string }
+): { [key: string]: string } {
+    const isNotOperator = (condition.operator as string) === 'not'
+
+    const conditionFixer = (
+        alias: string,
+        column: string,
+        filter: Filter,
+        operator: WherePredicateOperator['operator'],
+        parameters: { [key: string]: string }
+    ): { condition_params: any; params: any } => {
+        let condition_params: any = undefined
+        let params = parameters
+        switch (operator) {
+            case 'between':
+                condition_params = [alias, `:${column}_from`, `:${column}_to`]
+                params = {
+                    [column + '_from']: filter[column].value[0],
+                    [column + '_to']: filter[column].value[1],
+                }
+                break
+            case 'in':
+                condition_params = [alias, `:...${column}`]
+                break
+            default:
+                condition_params = [alias, `:${column}`]
+                break
+        }
+        return { condition_params, params }
+    }
+
+    const { condition_params, params } = conditionFixer(
+        alias,
+        column,
+        filter,
+        isNotOperator ? condition['condition']['operator'] : condition.operator,
+        parameters
+    )
+
+    if (isNotOperator) {
+        condition['condition']['parameters'] = condition_params
+    } else {
+        condition.parameters = condition_params
+    }
+
+    return params
+}
+
+// This function is used to fix the column alias when using relation, embedded or virtual properties
+function fixColumnAlias(
```

Comment:
same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names